### ISSUE TITLE
No error reported when an impl Trait doesn't met the requirement of an associated type

### ISSUE THREAD

**Issue by yshui** [rust-lang/rust#78893]
### Code

I tried this code:

```rust
trait A {
    type T: Iterator<Item = i32>;
}

trait B {
    fn x(_: impl A<T = impl Send>);
}
```

I expected to see this happen: 

```
error[E0277]: `impl Send` is not an iterator
```

Instead, this happened: 

Code compiles

### Version it worked on

It most recently worked on: 1.48.0-beta.8

### Version with regression

rust nightly 2020-11-08 (likely not the earliest)

**Comment by jonas-schievink** [comment]
cc @matthewjasper (might be caused by the projection bounds changes)

**Comment by camelid** [comment]
Note: not just `impl Trait`; also occurs with ordinary trait bounds.

**Comment by camelid** [comment]
searched nightlies: from nightly-2020-10-01 to nightly-2020-11-09
regressed nightly: nightly-2020-10-07
searched commits: from https://github.com/rust-lang/rust/commit/a1dfd2490a6cb456b92e469fa550dc217e20ad6d to https://github.com/rust-lang/rust/commit/98edd1fbf8a68977a2a7c1312eb1ebff80515a92
regressed commit: https://github.com/rust-lang/rust/commit/08e2d4616613716362b4b49980ff303f2b9ae654

<details>
<summary>bisected with <a href='https://github.com/rust-lang/cargo-bisect-rustc'>cargo-bisect-rustc</a> v0.6.0</summary>


Host triple: x86_64-apple-darwin
Reproduce with:
```bash
cargo bisect-rustc --preserve --regress=success --start=2020-10-01 
```
</details>

---

Indeed caused by #73905.

**Comment by camelid** [comment]
Assigning `P-critical` and removing `I-prioritize` as [discussed](https://rust-lang.zulipchat.com/#narrow/stream/245100-t-compiler.2Fwg-prioritization.2Falerts/topic/I-prioritize.20.2378893.20No.20error.20reported.20when.20an.20impl.20Trait.20do.E2.80.A6) in the prioritization working group.

**Comment by pnkfelix** [comment]
We discussed this in today's T-compiler triage meeting.

This may or may not represent a significant change to the static semantics. It may or may not represent a footgun.

My original instinct was to revert PR #73905 before the beta cut, to buy us time. But team members convinced me that would be rash, given the amount of work that builds upon PR #73905.

We decided it would be best to let PR #73905 remain landed for now, but we definitely want T-lang to double-check that we are all okay with the change to the static semantics here.

**Comment by oli-obk** [comment]
Here's the zulip thread if you want to join the discussion: https://rust-lang.zulipchat.com/#narrow/stream/131828-t-compiler/topic/Implied.20bounds.20analysis

**Comment by yshui** [comment]
@pnkfelix Interesting. Keeping this regression means the user wouldn't get an error until they actually use the type. To me this doesn't feel very desirable. But I can also see the argument that this is not a very big problem, because the user is likely going to use it, otherwise they wouldn't have written it out.

Perhaps this would break the pattern where the user defines a trait, uses the trait, but only implement the trait much later?

**Comment by oli-obk** [comment]
You can in fact implement both `A` and `B` in this situation (as the examples linked in the zulip show). All you did with `B` is to add additional constraints to callers of `B::x`. Though it is true that you don't have access (yet) to the bounds on `A::T` within `B::x`

**Comment by nikomatsakis** [comment]
Related test:

```rust
trait A {
    type T: Iterator<Item = i32>;
}

fn foo<X, Y: Send>() 
where
    X: A<T = Y>
{ 
}
```

this is an error on stable, but not nightly.

**Comment by nikomatsakis** [comment]
I did a little investigation here. I think that, for the second test I gave at least, the fix I would expect to do would be around here:

https://github.com/rust-lang/rust/blob/fe982319aa0aa5bbfc2795791a753832292bd2ba/compiler/rustc_trait_selection/src/traits/wf.rs#L117-L120

In particular, given a where clause like `X: A<T = Y>`, the first line checks that `<X as A>::T` is valid, the second checks that `Y` is valid, but nothing checks that `Y` meets the bounds declared for `T`. I believe that what we want is some code that fetches the `item_bounds` for `t.projection_ty.item_def_id`, applies the substitutions, and then replaces the "self" types with `Y`. I started tinkering with writing that code but it's mildly annoying, in part because the self types on `item_bounds` are full projection types -- I feel like it'd be nice if they were "bound" instead for easy substitution. (I'm not clear on whether `item_bounds` *only* applies to associated types?)

In any case, this won't fix the original example though. That would require a change to the opaque type rules:

**Comment by nikomatsakis** [comment]
Ok so the `impl Trait` WF code is here: 

https://github.com/rust-lang/rust/blob/fe982319aa0aa5bbfc2795791a753832292bd2ba/compiler/rustc_trait_selection/src/traits/wf.rs#L609-L613

Those rules currently fall through to the `nominal_obligations` here:

https://github.com/rust-lang/rust/blob/fe982319aa0aa5bbfc2795791a753832292bd2ba/compiler/rustc_trait_selection/src/traits/wf.rs#L677-L681

which invokes `predicates_of` on the opaque type. 

Or, well, hmm, maybe the right fix *there* is to modify how `wfcheck` applies to the opaque type definition that we synthesize. So one possible fix might be to make the previous change I mentioned and then extend

https://github.com/rust-lang/rust/blob/fe982319aa0aa5bbfc2795791a753832292bd2ba/compiler/rustc_typeck/src/check/wfcheck.rs#L78

with code to handle [`HirKind::OpaqueTy`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_hir/enum.ItemKind.html#variant.OpaqueTy) to check that the where clauses are well-formed (in the same way as we check for other items).

**Comment by Mark-Simulacrum** [comment]
This is likely going to be a stable/stable regression with the release of 1.49, but as it expands the set of compiling code I do not see it as a release blocker at this time (and it seems that a fix is at best unlikely in that time frame).

**Comment by yshui** [comment]
@Mark-Simulacrum if this gets into stable, the eventual fix of this bug is going to be a breaking change.

**Comment by pnkfelix** [comment]
This is a stable-to-stable regression now with the 1.49 release.

**Comment by chrysn** [comment]
Closing #80821 as a duplicate of this; the code provided there is something that can easily come up in development on 1.49, and all the types involved are used in the example code (showing that this is not just about errors in traits that can't be implemented anyway).

**Comment by jackh726** [comment]
This compiles now, marking as needs-test

**Comment by chrysn** [comment]
This has compiled all the way since 1.49, which is the problem: it should not have, the regression is that builds started to work. (AIU there is work going on to make requirements propagate implicitly, but these should still be behind a feature gate and were not part of release announcements that'd make people with MSRV requirements aware of them).

**Comment by jackh726** [comment]
Ah, I missed that. Relabeled accordingly.

**Comment by lcnr** [comment]
discussed during [t-types backlog bonanza](https://rust-lang.zulipchat.com/#narrow/stream/326132-t-types.2Fmeetings/topic/meeting.202022-07-15/near/289725758).

Going to discuss this in a later meeting to make a final discussion here

**Comment by jackh726** [comment]
Discussed today during [t-types meeting](https://rust-lang.zulipchat.com/#narrow/stream/326132-t-types.2Fmeetings/topic/meeting.202022-07-29/near/291317862).

We agreed that this should compile, given this also compiles:
```rust
trait B {
    fn x(_: impl A);
}
```
In other words, the bounds specified in the function arg are a convenient way to add new bounds, but don't change the fact that you still must prove the bounds on the trait.

We think this can be closed with the following tests (or test with revisions):
1) A test added showing this compiles
2) A test added that shows you can't call `x` without proving that `T: Iterator<Item = i32>`

**Comment by yshui** [comment]
@jackh726 Then shouldn't this also compile with the same logic?

```rust
trait A<T: Iterator<Item = i32>> {
}
trait B {
    fn x<T: Send>(_: impl A<T>);
}
```

Why is associated type treated differently from type parameters?

**Comment by lcnr** [comment]
@yshui we've ended up quickly talking about your question in the [t-types meeting](https://rust-lang.zulipchat.com/#narrow/stream/326132-t-types.2Fmeetings/topic/meeting.202022-07-29/near/291329478) as well

my intuition for this is that bounds on associated types have to be proven by the impl while bounds on generic arguments have to be proven by the user of that impl.

A bound like `T: Trait<Assoc = U>` means that `T: Trait`, and `<T as Trait>::Assoc == U` must hold.

The equality implies that all bounds on `<T as Trait>::Assoc` must also hold for `U` (as these two are equal), but it never requires it. Both `T: Trait` and `<T as Trait>::Assoc == U` do not require `U: Iterator<Item = i32>`.

If you have `T: Trait<U>`, you cannot destruct this into smaller goals, so you have to prove all the requirements for this, including `U: Iterator<Item = i32>`.

As mentioned in the meeting itself, this isn't too great of an explanation if you aren't already deeply involved in the trait system, but I hope it was at least somewhat understandable :sweat_smile:

**Comment by yshui** [comment]
@lcnr This seems to be tied to the specific implementation of the trait system in rustc. And, at least to me, doesn't feel like an intuitive distinction between the 2 cases.

i.e. another implementation of the Rust language might come along and implement this differently. 

But I guess in the end this isn't that impactful of a change, and I can go with the decision. :+1:

**Comment by lcnr** [comment]
cc https://rust-lang.zulipchat.com/#narrow/channel/144729-t-types/topic/E-mentor.3A.20check.20item.20bounds.20for.20Projection.20where-clause.20WF/with/554691484

My opinion on this issue has changed since the types team meeting 3 years ago and I think we should try to change our internal implementation here. I still think implying these bounds is fine from a user-standpoint, but think not doing so in the implementation itself is causing a lot of issues
